Pep8 action (ruff)

PodushkaPIR opened the pull request:
- Add base ruff action in CI to catch pep8 suggestions (can be extended)
- Add ruff-tool in the pyproject.toml (can be extend)

This PR will change depending on the offers. I'm not sure which Errors and Warnings are better to introduce and which are better to ignore. I am ready for suggestions.

Also ruff is a great tool and and some features from it can be transferred to the project.

PodushkaPIR committed:
Add ruff-action to check python codestyle

PodushkaPIR committed:
Fix pyfiles with unused imports and an extra formatting print

PodushkaPIR committed:
Replace lambda to def with E731

PodushkaPIR committed:
Add ruff linter configuration with comprehensive rule groups

- Enable E, W, F, Q, T, D, S, UP, ISC, ICN rule groups
- Add line-length = 120
- Add ignore rules with comments explaining why each is ignored
- Add per-file-ignores for examples

PodushkaPIR committed:
Remove from blanklines the extra whitespaces

PodushkaPIR committed:
Fix invalid escape sequence W605

wildsor reviewed (changes requested):
the ruff-check rename should be squashed with the initial commit

BUYT-1 commented on pyproject.toml:
```diff
@@ -49,3 +49,19 @@ exclude = [
+
+[tool.ruff]
+select = [
+    "E", "W", "F",
```
Why not enable ALL and exclude stuff that doesn't make sense?

wildsor reviewed (changes requested):
I know that's not how they are written now, but I'd like CI config files to be written so that anyone looking at them for the first time can at least somewhat understand what's going on with minimal disruptions.
At least, if you're disabling or enabling something, there should be a rationale: if I just look at `[tool.ruff] ignore = [ "W291", ...]`, there is no way I can guess what it is doing without going into ruff's docs.

PodushkaPIR committed:
Fix trailing whitespace in Python files

PodushkaPIR committed:
Fix UP rules (pyupgrade)

- Convert .format() calls to f-strings (UP032)
- Replace %d format specifier with f-string (UP031)
- Remove unnecessary UTF-8 encoding declaration (UP009)
- Remove unnecessary __future__ import unicode_literals (UP010)
- Remove unnecessary mode='r' argument from open() (UP015)

PodushkaPIR committed:
Fix ISC rules (string concatenation)

- Remove redundant '+' operator for implicit string concatenation (ISC003)
- Fix implicit string concatenation over multiple lines (ISC002)
- Also fix import conventions (ICN001): pandas -> pd

PodushkaPIR committed:
Fix I001 import sorting in examples and test files

- Sort imports according to isort rules
- Group standard library, third-party, and local imports properly
- Also fix import conventions (ICN001): pandas -> pd

PodushkaPIR committed:
Fix snapshottest version for Python 3.12+ compatibility

Upgrade snapshottest from 1.0.0a1 to 0.6.0 for Python 3.12+ compatibility.
Old version uses deprecated `imp` module which was removed in Python 3.12.
Version 0.6.0 uses modern `importlib` instead.

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+    "ICN", # use common import conventions
+]
+# Ignore rules that are too strict for examples/notebooks:
+# - Q000: requires double quotes (examples use single quotes for consistency)
+# - Q001: single quote multiline strings (used for output formatting)
+# - Q002: move single-quoted strings to double (related to Q001)
+# - Q003: change outer quotes to avoid escaping (examples use escaped quotes)
```
These ones prevent inconsistencies. Let's enable them.

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+# - D107: missing docstring in __init__ (simple init methods)
+# - D203: one blank line required before class docstring
+#         (conflicts with D211: no blank line required before class)
+# - D210: one-line docstring should fit on first line (formatting preference)
```
The summary is incorrect

```suggestion
# - D210: No whitespaces allowed surrounding docstring text
```

I think we should enable this

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+# - D210: one-line docstring should fit on first line (formatting preference)
+# - D212: summary must start at first line of multiline docstring
+#         (conflicts with D213: summary must start at second line)
+# - D300: use triple double quotes (benchmark uses single quotes)
```
Let's fix it, PEP257 recommends double quotes for docstrings (consistency)

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+# - D212: summary must start at first line of multiline docstring
+#         (conflicts with D213: summary must start at second line)
+# - D300: use triple double quotes (benchmark uses single quotes)
+# - D400: one-line docstring does not match summary (formatting preference)
```
```suggestion
# - D400: First line should end with a period (formatting preference)
```
Let's fix this one

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+    "ISC", # good use of string concatenation
+    "ICN", # use common import conventions
+]
+# Ignore rules that are too strict for examples/notebooks:
+# - Q000: requires double quotes (examples use single quotes for consistency)
+# - Q001: single quote multiline strings (used for output formatting)
+# - Q002: move single-quoted strings to double (related to Q001)
+# - Q003: change outer quotes to avoid escaping (examples use escaped quotes)
+# - T201: warns about print() (examples use print for output demonstration)
+# - D100: missing docstring in public module (examples are scripts, not libraries)
+# - D101: missing docstring in public class (simple example classes)
+# - D102: missing docstring in public method (example methods prioritize clarity)
+# - D103: missing docstring in public function (simple helper functions)
+# - D104: missing docstring in public nested (helper functions in examples)
+# - D107: missing docstring in __init__ (simple init methods)
+# - D203: one blank line required before class docstring
+#         (conflicts with D211: no blank line required before class)
+# - D210: one-line docstring should fit on first line (formatting preference)
+# - D212: summary must start at first line of multiline docstring
+#         (conflicts with D213: summary must start at second line)
+# - D300: use triple double quotes (benchmark uses single quotes)
+# - D400: one-line docstring does not match summary (formatting preference)
+# - D413: bad multiline summary (formatting preference)
+# - D415: section underline (formatting preference)
+# - S101: use of assert detected (examples use assert for test demonstrations)
+# - S603: use of subprocess call (not a security issue in examples)
```
https://docs.astral.sh/ruff/rules/ has short descriptions for the names, you can use them

PodushkaPIR replied:
thanks

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+#         (conflicts with D213: summary must start at second line)
+# - D300: use triple double quotes (benchmark uses single quotes)
+# - D400: one-line docstring does not match summary (formatting preference)
+# - D413: bad multiline summary (formatting preference)
```
```suggestion
# - D413: Missing blank line after last section (formatting preference)
```

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+# - D300: use triple double quotes (benchmark uses single quotes)
+# - D400: one-line docstring does not match summary (formatting preference)
+# - D413: bad multiline summary (formatting preference)
+# - D415: section underline (formatting preference)
```
```suggestion
# - D415: First line should end with a period, question mark, or exclamation point
```
Let's fix this

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+
+[tool.ruff]
+# line-length: 120 instead of default 88
+# Reason: Examples often have long f-strings with column names, URLs, table data
```
https://docs.astral.sh/ruff/rules/line-too-long/:
> In the interest of pragmatism, this rule makes a few exceptions when determining whether a line is overlong. Namely, it:
> * Ignores lines that consist of a single "word" (i.e., without any whitespace between its characters).
> * Ignores lines that end with a URL, as long as the URL starts before the line-length threshold.
> * Ignores line that end with a pragma comment (e.g., # type: ignore or # noqa), as long as the pragma comment starts before the line-length threshold. That is, a line will not be flagged as overlong if a pragma comment causes it to exceed the line length. (This behavior aligns with that of the Ruff formatter.)
> * Ignores SPDX license identifiers and copyright notices (e.g., # SPDX-License-Identifier: MIT), which are machine-readable and should not wrap over multiple lines.

Where are the false positives? I wonder if it would be better to explicitly ignore exceptions.

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+
+[tool.ruff.lint.per-file-ignores]
+# E702: Multiple statements on one line (e.g., G.add_node(); G.add_edge())
+"examples/**/*.py" = ["E501", "E702"]
```
Let's fix E702, I think it's better to leave awkward code in, it lets us see awkward interfaces

wildsor commented on pyproject.toml:
```diff
@@ -131,3 +131,70 @@ extend-ignore-re = [
+# E501: Long lines in cell outputs
+# E731: lambda assigned to variable (cell-specific pattern)
+# F401, F811, F821: Unused imports/redefinitions (cell scoping)
+"examples/notebooks/*.ipynb" = ["E501", "E731", "F401", "F811", "F821"]
```
How does it break? Does it actually check every cell individually? That's strange, I wonder if there is a setting for it
